% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/scoreMotif.R
\name{plotMB}
\alias{plotMB}
\title{Plot a genomic region surrounding a genomic variant, and potentially disrupted
motifs}
\usage{
plotMB(results, rsid, reverseMotif = TRUE, effect = c("strong", "weak"))
}
\arguments{
\item{results}{The output of \code{motifbreakR}}

\item{rsid}{Character; the identifier of the variant to be visualized}

\item{reverseMotif}{Logical; if the motif is on the "-" strand show the
the motifs as reversed \code{FALSE} or reverse complement \code{TRUE}}

\item{effect}{Character; show motifs that are strongly effected \code{c("strong")},
weakly effected \code{c("weak")}, or both \code{c("strong", "weak")}}
}
\value{
plots a figure representing the results of \code{motifbreakR} at the
  location of a single SNP, returns invisible \code{NULL}.
}
\description{
Plot a genomic region surrounding a genomic variant, and potentially disrupted
motifs
}
\details{
\code{plotMB} produces output showing the location of the SNP on the
  chromosome, the surrounding sequence of the + strand, the footprint of any
  motif that is disrupted by the SNP or SNV, and the DNA sequence motif(s)
}
\examples{
data(example.results)
example.results
\dontrun{
plotMB(example.results, "rs2661839", effect = "strong")
}
}
\seealso{
See \code{\link{motifbreakR}} for the function that produces output to be
  visualized here, also \code{\link{snps.from.rsid}} and \code{\link{snps.from.file}}
  for information about how to generate the input to \code{\link{motifbreakR}}
  function.
}
